### PR TITLE
Show ibus methods

### PR DESCRIPTION
**Pull request by jeremypw**
Needs https://github.com/elementary/switchboard-plug-keyboard/pull/341  (OR manual entry of an ibus method into org.gnome.desktop.input-source.sources) to have an effect.

Also requires https://github.com/elementary/gala/pull/1041 to be installed to have selected IBus method use the expected keyboard layout.

### CODE REVIEW

**Comment by jeremypw** [comment]
Input source Xkb keyboard (UK English layout);
Screenshot from 2021-02-05 10 18 01

Input source Ibus Mozc Japanese using UK English keyboard layout
Screenshot from 2021-02-05 10 20 53

**Comment by ryonakano** [comment]
> Input source Xkb keyboard (UK English layout);
> Screenshot from 2021-02-05 10 18 01
> 
> Input source Ibus Mozc Japanese using UK English keyboard layout
> Screenshot from 2021-02-05 10 20 53

It might be a little strange that we can check two radiobuttons here (xkb and ibus engine) :thinking:

**Comment by jeremypw** [comment]
> It might be a little strange that we can check two radiobuttons here (xkb and ibus engine) :thinking:

How about if there is a separator between the xkb and ibus lists?

**Comment by jeremypw** [comment]
With separator:
image

**Comment by jeremypw** [comment]
A couple questions remain:

1) Do we need any headers to indicate what the two lists denote?
2) It is not immediately obbious that to turn off an ibus input method you click on a keyboard layout (even if it is already ticked).  Do we need a "None" entry in the input method list (when input methods are available) or a switch?

**Comment by danrabbit** [comment]
1. I'm not certain a separator makes things clear on its own here; header labels make sense to me.
2. As long as that works conceptually in terms of how you think about input methods (ie: its clear that Ibus is a service you are enabling or disabling) a switch with the h4 style class could work and also act as the list header. The list of input methods could hide when it's turned off.

**Comment by jeremypw** [comment]
After latest commits:

Screenshot from 2021-03-14 17 08 25

**Comment by jeremypw** [comment]
The necessary Gala and Greeter fixes have now been merged.

File: src/LayoutsManager.vala [hunk 1]
```diff
@@ -16,77 +16,231 @@
...
+    private IBus.Bus bus;
+    private SimpleActionGroup actions;
 
     construct {
-        main_grid = new Gtk.Grid () {
+        IBus.init ();
+        bus = new IBus.Bus ();
+
+        xkb_header = new Gtk.Label (_("Keyboard layout")) {
+            margin = 3,
```
Comment (danrabbit) [inline src/LayoutsManager.vala:50]: Can we take this out? I think it would be better to set the margins in the stylesheet

File: src/LayoutsManager.vala [hunk 2]
```diff
@@ -16,77 +16,231 @@
...
+
+    private IBus.Bus bus;
+    private SimpleActionGroup actions;
 
     construct {
-        main_grid = new Gtk.Grid () {
+        IBus.init ();
+        bus = new IBus.Bus ();
+
+        xkb_header = new Gtk.Label (_("Keyboard layout")) {
```
Comment (danrabbit) [inline src/LayoutsManager.vala:49]: Headers should be title case :)
```suggestion
        xkb_header = new Gtk.Label (_("Keyboard Layout")) {
```

File: src/LayoutButton.vala [hunk 3]
```diff
@@ -15,39 +15,52 @@
...
+            action_name = _action_name,
+            action_target = _action_target,
+            text = caption
+        };
 
-        modelbutton.button_release_event.connect (() => {
-            settings.set_value ("current", id);
-            return Gdk.EVENT_STOP;
-        });
+        button.get_style_context ().add_class (Gtk.STYLE_CLASS_RADIO);
```
Comment (danrabbit) [inline src/LayoutButton.vala:60]: Does this do the same as https://valadoc.org/gtk+-3.0/Gtk.ModelButton.role.html ?

File: src/LayoutsManager.vala [hunk 4]
```diff
@@ -16,77 +16,231 @@
...
+        var ibus_header_grid = new Gtk.Grid () {
+            orientation = Gtk.Orientation.VERTICAL
+        };
+        var ibus_separator = new Gtk.Separator (Gtk.Orientation.HORIZONTAL);
+        ibus_header = new Granite.SwitchModelButton (_("Input Method")) {
+            active = true
+        };
+        ibus_header.get_style_context ().add_class (Granite.STYLE_CLASS_H4_LABEL);
+        ibus_header_grid.add (ibus_separator);
+        ibus_header_grid.add (ibus_header);
```
Comment (danrabbit) [inline src/LayoutsManager.vala:71]: It's a little hard to follow what's going on with things jumbled and no white space:
```suggestion
        var ibus_separator = new Gtk.Separator (Gtk.Orientation.HORIZONTAL);

        ibus_header = new Granite.SwitchModelButton (_("Input Method")) {
            active = true
        };
        ibus_header.get_style_context ().add_class (Granite.STYLE_CLASS_H4_LABEL);

        var ibus_header_grid = new Gtk.Grid () {
            orientation = Gtk.Orientation.VERTICAL
        };
        ibus_header_grid.add (ibus_separator);
        ibus_header_grid.add (ibus_header);
```

File: src/LayoutsManager.vala [hunk 5]
```diff
@@ -16,77 +16,231 @@
...
+        });
+
+        ibus_grid.get_children ().foreach ((child) => {
+            child.destroy ();
+        });
+
+        ibus_header_revealer.reveal_child = false;
+        ibus_grid_revealer.reveal_child = false;
+
+
```
Comment (danrabbit) [inline src/LayoutsManager.vala:138]: double whitespace

File: src/LayoutsManager.vala [hunk 6]
```diff
@@ -16,77 +16,231 @@
  */
 
 
-public class Keyboard.Widgets.LayoutManager : Gtk.ScrolledWindow {
+public class Keyboard.Widgets.LayoutManager : Gtk.ListBox {
```
Comment (danrabbit) [inline src/LayoutsManager.vala:19]: Why is this a listbox now?

**Review by danrabbit — CHANGES_REQUESTED** [review]
Just a few comments, but I think this makes sense from a design perspective

File: src/LayoutsManager.vala (hunk 6 above)
Comment (jeremypw) [reply src/LayoutsManager.vala:19]: I think it was partly because that is what other wingpanel indicators use and I thought it would make styling easier. The number of potential entries does not warrant a ScrolledWindow.

File: src/LayoutsManager.vala (hunk 1 above)
Comment (jeremypw) [reply src/LayoutsManager.vala:50]: Done

File: src/LayoutsManager.vala (hunk 2 above)
Comment (jeremypw) [reply src/LayoutsManager.vala:49]: Done

File: src/LayoutsManager.vala (hunk 4 above)
Comment (jeremypw) [reply src/LayoutsManager.vala:71]: Done

File: src/LayoutsManager.vala (hunk 5 above)
Comment (jeremypw) [reply src/LayoutsManager.vala:138]: Fixed

File: src/LayoutButton.vala (hunk 3 above)
Comment (jeremypw) [reply src/LayoutButton.vala:60]: Yes, in fact the appearance is determined from the action automatically so I will take out this line.

File: src/LayoutsManager.vala (hunk 6 above)
Comment (danrabbit) [reply src/LayoutsManager.vala:19]: As far as I can tell, the only things that get packed here are the revealers for the header and the actual list of options right? So shouldn't this just be a grid?

**Comment by jeremypw** [comment]
@danrabbit Will you fix the header margin issue in the stylesheet or does something need to be done in this PR?

Based on Gtk.Grid

Screenshot from 2021-03-17 18 48 46

**Comment by danrabbit** [comment]
Yep I'll take care of that in the system stylesheet :)

**Review by danrabbit — APPROVED** [review]
I'm happy if @ryonakano is happy :)

**Comment by jeremypw** [comment]
One thing occurred to me - do we need to include code to detect pre-existing ibus input methods which were set up before the new version of the switchboard plug and there not integrated into the `input-sources` setting?  Opening and closing the switchboard keyboard plug is required update the setting.

**Comment by jeremypw** [comment]
I think it best to merge this now that the corresponding switchboard plug PR is merged, and deal with any upgrade problems later.
